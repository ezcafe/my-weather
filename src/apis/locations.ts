import HttpErrorResponseModel from '../models/HttpErrorResponseModel';
import * as ApiUtility from '../utilities/ApiUtility';
import LocationModel from '../stores/locations/models/LocationModel';
import LocationWeatherModel from '../stores/locations/models/LocationWeatherModel';

export async function fetchLocations(
    location: string
): Promise<LocationModel[] | HttpErrorResponseModel> {
    const endpoint: string = '/api/location/search/';
    return ApiUtility.getToModel<LocationModel[]>(LocationModel, endpoint, {
        query: location,
    });
}

export async function fetchLocationWeather(
    woeid: string
): Promise<LocationWeatherModel | HttpErrorResponseModel> {
    const endpoint: string = `/api/location/${woeid}/`;
    return ApiUtility.getToModel<LocationWeatherModel>(
        LocationWeatherModel,
        endpoint
    );
}
