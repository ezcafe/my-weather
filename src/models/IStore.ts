import { RouterState } from 'connected-react-router';
import IErrorState from '../stores/error/models/IErrorState';
import ILocationsState from '../stores/locations/models/ILocationsState';
import IRequestingState from '../stores/requesting/models/IRequestingState';
import IToastsState from '../stores/toasts/models/IToastsState';

export default interface IStore {
    readonly error: IErrorState;
    readonly requesting: IRequestingState;
    readonly router: RouterState;
    readonly locations: ILocationsState;
    readonly toasts: IToastsState;
}
