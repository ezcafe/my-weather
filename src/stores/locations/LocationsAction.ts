import LocationModel from './models/LocationModel';
import LocationWeatherModel from './models/LocationWeatherModel';
import HttpErrorResponseModel from '../../models/HttpErrorResponseModel';
import IAction from '../../models/IAction';
import * as ActionUtility from '../../utilities/ActionUtility';

export const REQUEST_LOCATIONS: string = 'LocationsAction.REQUEST_LOCATIONS';
export const REQUEST_LOCATIONS_FINISHED: string =
    'LocationsAction.REQUEST_LOCATIONS_FINISHED';

export const REQUEST_LOCATION_WEATHER: string =
    'LocationsAction.REQUEST_LOCATION_WEATHER';
export const REQUEST_LOCATION_WEATHER_FINISHED: string =
    'LocationsAction.REQUEST_LOCATION_WEATHER_FINISHED';

export function requestLocationsAction(location: string): IAction<string> {
    return ActionUtility.createAction(REQUEST_LOCATIONS, location);
}

export function requestLocationsFinishedAction(
    model: LocationModel[] | HttpErrorResponseModel
): IAction<LocationModel[] | HttpErrorResponseModel> {
    const isError: boolean = model instanceof HttpErrorResponseModel;
    return ActionUtility.createAction(
        REQUEST_LOCATIONS_FINISHED,
        model,
        isError
    );
}

export function requestLocationWeatherAction(woeid: string): IAction<string> {
    return ActionUtility.createAction(REQUEST_LOCATION_WEATHER, woeid);
}

export function requestLocationWeatherFinishedAction(
    model: LocationWeatherModel | HttpErrorResponseModel
): IAction<LocationWeatherModel | HttpErrorResponseModel> {
    const isError: boolean = model instanceof HttpErrorResponseModel;
    return ActionUtility.createAction(
        REQUEST_LOCATION_WEATHER_FINISHED,
        model,
        isError
    );
}
