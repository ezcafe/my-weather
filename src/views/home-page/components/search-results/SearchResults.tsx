import { useSelector } from 'react-redux';
import SearchResult from './SearchResult';
import LoadingIndicator from '../../../components/loading-indicator/LoadingIndicator';
import {
    locationWeatherSelector,
    isLoadingLocationWeatherSelector,
} from '../../../../stores/locations/LocationsSelector';
import LocationWeatherModel from '../../../../stores/locations/models/LocationWeatherModel';
import LocationConsolidatedWeatherModel from '../../../../stores/locations/models/LocationConsolidatedWeatherModel';

const SearchResults: React.FC = () => {
    const isLoadingLocationWeather: boolean = useSelector(
        isLoadingLocationWeatherSelector
    );
    const locationWeather: LocationWeatherModel | undefined = useSelector(
        locationWeatherSelector
    );

    if (isLoadingLocationWeather) {
        return <LoadingIndicator />;
    }

    if (!locationWeather) {
        return null;
    }

    const consolidatedWeather: LocationConsolidatedWeatherModel[] =
        locationWeather.consolidated_weather;
    consolidatedWeather.length = 5; // limit 5 dates

    return (
        <>
            {consolidatedWeather.map(
                (model: LocationConsolidatedWeatherModel) => (
                    <SearchResult key={model.id} item={model} />
                )
            )}
        </>
    );
};

export default SearchResults;
