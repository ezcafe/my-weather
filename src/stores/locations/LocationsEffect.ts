import { takeLatest, call, put } from 'redux-saga/effects';
import {
    REQUEST_LOCATIONS,
    REQUEST_LOCATION_WEATHER,
    requestLocationsFinishedAction,
    requestLocationWeatherFinishedAction,
} from './LocationsAction';
import LocationModel from './models/LocationModel';
import LocationWeatherModel from './models/LocationWeatherModel';
import { fetchLocations, fetchLocationWeather } from '../../apis/locations';
import IAction from '../../models/IAction';
import HttpErrorResponseModel from '../../models/HttpErrorResponseModel';

function* requestLocations(action: IAction<string>) {
    const { payload: location } = action;
    if (!location) {
        return;
    }
    const model: LocationModel[] | HttpErrorResponseModel = yield call(
        fetchLocations,
        location
    );
    yield put(requestLocationsFinishedAction(model));
}

function* requestLocationWeather(action: IAction<string>) {
    const { payload: woeid } = action;
    if (!woeid) {
        return;
    }
    const model: LocationWeatherModel | HttpErrorResponseModel = yield call(
        fetchLocationWeather,
        woeid
    );
    yield put(requestLocationWeatherFinishedAction(model));
}

export default function* locationsEffectWatcher() {
    yield takeLatest(REQUEST_LOCATIONS, requestLocations);
    yield takeLatest(REQUEST_LOCATION_WEATHER, requestLocationWeather);
}
