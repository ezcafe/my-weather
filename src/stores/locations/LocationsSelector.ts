import LocationModel from './models/LocationModel';
import LocationWeatherModel from './models/LocationWeatherModel';
import IStore from '../../models/IStore';

export function locationsSelector(state: IStore): LocationModel[] {
    return state.locations.items;
}

export function locationWeatherSelector(
    state: IStore
): LocationWeatherModel | undefined {
    return state.locations.weather;
}
