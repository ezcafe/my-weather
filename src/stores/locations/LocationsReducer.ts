import ILocationsState from './models/ILocationsState';
import * as LocationsAction from './LocationsAction';
import IAction from '../../models/IAction';
import LocationModel from './models/LocationModel';
import LocationWeatherModel from './models/LocationWeatherModel';
import baseReducer from '../../utilities/BaseReducer';
import { Reducer } from 'redux';

export const initialState: ILocationsState = {
    items: [],
    weather: undefined,
};

const locationsReducer: Reducer = baseReducer(initialState, {
    [LocationsAction.REQUEST_LOCATIONS](
        state: ILocationsState,
        action: IAction<string>
    ): ILocationsState {
        return {
            ...state,
            items: [], // reset items and show loading icon in search form
        };
    },
    [LocationsAction.REQUEST_LOCATIONS_FINISHED](
        state: ILocationsState,
        action: IAction<LocationModel[]>
    ): ILocationsState {
        return {
            ...state,
            items: action.payload!,
        };
    },
    [LocationsAction.REQUEST_LOCATION_WEATHER_FINISHED](
        state: ILocationsState,
        action: IAction<LocationWeatherModel>
    ): ILocationsState {
        return {
            ...state,
            weather: action.payload!,
        };
    },
});

export default locationsReducer;
