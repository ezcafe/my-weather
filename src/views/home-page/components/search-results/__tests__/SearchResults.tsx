import { render, RenderResult } from '@testing-library/react';
import SearchResults from '../SearchResults';
import { mockedLocations } from '../../../../../fixtures';
import withMockedRedux from '../../../../../fixtures/withMockedRedux';
import IStore from '../../../../../models/IStore';
import { REQUEST_LOCATION_WEATHER } from '../../../../../stores/locations/LocationsAction';

describe('SearchResults', () => {
    function renderComponent(initialState?: Partial<IStore>): RenderResult {
        const WrappedComponent = withMockedRedux(SearchResults, initialState);
        return render(<WrappedComponent />);
    }

    test('should show loading indicator', () => {
        const { container } = renderComponent({
            requesting: {
                [REQUEST_LOCATION_WEATHER]: true,
            },
        });
        expect(container).toMatchSnapshot();
    });

    test('should show nothing', () => {
        const { container } = renderComponent({
            locations: {
                items: [],
                weather: undefined,
            },
        });
        expect(container).toMatchSnapshot();
    });

    test('should show results', () => {
        const { container } = renderComponent({
            locations: mockedLocations,
        });
        expect(container).toMatchSnapshot();
    });
});
