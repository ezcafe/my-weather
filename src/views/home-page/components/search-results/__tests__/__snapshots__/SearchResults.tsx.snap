// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults should show loading indicator 1`] = `
<div>
  <div
    class="ai spinner-border spinner-border-sm"
    role="status"
  >
    <span
      class="sr-only"
    >
      Loading...
    </span>
  </div>
</div>
`;

exports[`SearchResults should show nothing 1`] = `<div />`;

exports[`SearchResults should show results 1`] = `
<div>
  <div
    class="col"
  >
    <div
      class="search-result"
    >
      <img
        alt="Heavy Cloud"
        src="https://www.metaweather.com/static/img/weather/hc.svg"
      />
      <h2>
        Wednesday
      </h2>
      <span>
        Min: 8
      </span>
      <span>
        Max: 12
      </span>
    </div>
  </div>
</div>
`;
